utilities/blob_db/blob_db_impl.cc

```diff
@@ -0,0 +1,2073 @@
+std::shared_ptr<blob_log::Reader> BlobFile::openSequentialReader(
+  Env *env, const DBOptions& db_options, const EnvOptions& env_options) const {
+  std::unique_ptr<SequentialFile> sfile;
+  Status s = env->NewSequentialFile(PathName(), &sfile, env_options);
+  if (!s.ok()) {
+    // report something here.
+    return nullptr;
+  }
+
+  std::unique_ptr<SequentialFileReader> sfile_reader;
+  sfile_reader.reset(new SequentialFileReader(std::move(sfile)));
+
+  std::shared_ptr<blob_log::Reader> log_reader =
+    std::make_shared<blob_log::Reader>(db_options.info_log,
+    std::move(sfile_reader), nullptr, true, 0, file_number_);
+
+  return log_reader;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+//
+//
+////////////////////////////////////////////////////////////////////////////////
+bool BlobFile::NeedsFsync(bool hard, uint64_t bytes_per_sync) const {
+  assert (last_fsync_ <= file_size_);
+  return (hard) ? file_size_ > last_fsync_ :
+    (file_size_ - last_fsync_) >= bytes_per_sync;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+//
+//
+////////////////////////////////////////////////////////////////////////////////
+Status BlobFile::writeFooterAndClose_locked()
+{
+  Log(InfoLogLevel::INFO_LEVEL, parent_->db_options_.info_log,
+    "File is being closed after footer %s", PathName().c_str());
+
+  blob_log::BlobLogFooter footer;
+  footer.blob_count_ = blob_count_;
+  footer.ttl_range_ = ttl_range_;
+  footer.sn_range_ = sn_range_;
+  footer.ts_range_ = time_range_;
+
+  footer.has_ttl_ = HasTTL();
+  footer.has_ts_ = HasTimestamps();
+
+  // this will close the file and reset the Writable File Pointer.
+  Status s = log_writer_->AppendFooter(footer);
+  if (s.ok()) {
+    closed_ = true;
+    file_size_ += blob_log::BlobLogFooter::kFooterSize;
+  } else {
+    Log(InfoLogLevel::ERROR_LEVEL, parent_->db_options_.info_log,
+      "Failure to read Header for blob-file %s", PathName().c_str());
+  }
+  // delete the sequential writer
+  log_writer_.reset();
+  return s;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+//
+//
+////////////////////////////////////////////////////////////////////////////////
+void BlobFile::setHeader_locked(const blob_log::BlobLogHeader& hdr) {
+  header_ = hdr;
+  header_valid_ = true;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+//
+//
+////////////////////////////////////////////////////////////////////////////////
+void BlobFile::setFromFooter_locked(const blob_log::BlobLogFooter& footer)
+{
+  blob_log::BlobLogHeader hdr;
+  hdr.setTTL(footer.HasTTL());
+  hdr.setTimestamps(footer.HasTimestamps());
+  setHeader_locked(hdr);
+
+  // assume that file has been fully fsync'd
+  last_fsync_.store(file_size_);
+  blob_count_ = footer.GetBlobCount();
+  ttl_range_ = footer.GetTTLRange();
+  time_range_ = footer.GetTimeRange();
+  sn_range_ =
```

Comment:
moved out BlobFile into separate file. BlobDBImpl is still the majority of the code, unfortunately and maybe if it is still very big, can split it someway.